Bluetooth: controller: Next Generation Scheduler [framework]

mtpr-ot opened the pull request:
This commit splits code from Nordic LLL into lll_common.c, which can
then be reused across LLL implementations.

The ticker can now be configured in a "slot agnostic" mode, in which no
collision resolving or slot reservation is used. In this mode, the
ticker acts as a simple timer.

Event scheduling states are stored in the lll_hdr, and event priority
is passed from LLL implementation, and runtime priority calculated.

LLL implementation decides whether to program radio, start preemption
timer, and/or queue prepare in the prepare pipeline.

Event arbitration is made possible via the common LLL, but not yet
implemented in Nordic LLL.

Split only covers Nordic implementation. Other vendors should adjust
and implement their LLL accordingly to use the feature.

Signed-off-by: Morten Priess <mtpr@oticon.com>

mtpr-ot commented:
This is a first, barebone implementation which provides the necessary framework for implementing a new scheduling paradigm.
The commit is not intended for merge in it's current form, but as a basis for discussions regarding concept and naming etc.

@cvinayak, I would very much like if you could take a look at the PR and comment, question, anything.

FYI, with this concept and PR, Oticon LLL has been reimplemented with far more efficient event scheduling, more concurrent stable connections and higher data throughput. The possibility of soon supporting ISO pause/resume is also inherent in the design.

Notice that ticker is almost back to basics, and if this concept becomes mainstream, we should probably strip down the module.

cvinayak commented on subsys/bluetooth/controller/ll_sw/lll.h:
```diff
@@ -155,18 +155,26 @@ struct ull_hdr {
 typedef int (*lll_prepare_cb_t)(struct lll_prepare_param *prepare_param);
-typedef int (*lll_is_abort_cb_t)(void *next, int prio, void *curr,
-				 lll_prepare_cb_t *resume_cb, int *resume_prio);
+typedef int (*lll_is_abort_cb_t)(void *next, void *curr,
+				 lll_prepare_cb_t *resume_cb);
```
A separate commit for refactoring/removal of `prio` and `resume_prio`.

cvinayak reviewed (changes requested):
Summary of changes (separate commits):
1. Rename `BT_TICKER_COMPATIBILITY_MODE` to `BT_TICKER_LOW_LAT`
2. Introduction of `score`
3. Introduction of `latency`
4. Removal of `prio` and `resume_prio`
5. Introduction of `force` in the ticker callback interface

The direction this is heading looks good to me.

mtpr-ot commented on subsys/bluetooth/controller/Kconfig.ll_sw_split:
```diff
@@ -528,14 +528,32 @@ config BT_TICKER_COMPATIBILITY_MODE
 config BT_TICKER_EXT
 	bool "Ticker extensions"
-	depends on !BT_TICKER_COMPATIBILITY_MODE
+	depends on !BT_TICKER_COMPATIBILITY_MODE && !BT_TICKER_SLOT_AGNOSTIC
```
Good idea with the rename.
> 2. Is slot_window still supported?

No, this is something that needs to be reworked in the ull_adv (and ull_scan if needed). Since there is no notion of slot reservation and collision, the slot_window cannot be used to reschedule within a window. For ADV, it's reasonably simple to implement the same "windowing" mechanics in LL, if it does not get to run. I have a solution for that.

cvinayak replied:
What about a replacement for `ticker_next_slot_get`, this is something currently used to fill Connection Parameter Request offsets, schedule non-overlapping central  connections, Auxiliary Pointer offset, Sync Info offset, BIGInfo offset etc... Do you plan to supply/share the time reservation/usage by each role in the new framework?

mtpr-ot replied:
> What about a replacement for `ticker_next_slot_get`

Yes, this is a very good question. I am aware of the `ticker_next_slot_get` functionality. Even though I must say I don't like it very much, I agree it is needed for the purposes you mention.
So how to handle this without the slot reservation? Well, the function only uses the slot for finding ticker nodes that care about "air-time" - it doesn't use the slot reservation as such. It should be a small change to have it search for maybe a specific ticker node id or range. The result is anyway checked against ticker id and disqualified in some cases, causing a new iteration of searching. The best solution is of course to have the searching find the right ticker in the first attempt, and that can probably be implemented by passing a "search criterion".
The need for locating optimal ticker timing "locations" is growing. Vendors will be needing to request the best possible location for placing central connections, to allow space for large audio packets to two or more peripherals.
But tell me, did I miss some important fact here?

cvinayak replied:
>  Vendors will be needing to request the best possible location for placing central connections, to allow space for large audio packets to two or more peripherals.

`ticker_next_slot_get` is an enumeration interface, in a primitive form which is used currently to place (for now central, but shall be used to group auxiliary and periodic PDUs too).

The ticker concept is to keep time reservations in a central location (other concept is, distributed in users of ticker, in advertisers/scanner/central/peripheral etc., needing a framework across users adding to a variable execution time dependencies based on how a role is decided to be implemented in the framework), such that its scalable with respect to users. These time reservations shall be the bare minimal to guarantee the minimal CPU/Radio and other resource availability.

Same applies to priorities too, keep it centralized (as in real hardware SoC's, interrupt controllers). `score` concept seems inline towards fairness when `abort` or `too late` happens in LLL pipeline. I do not like distributed dynamic priorities, but weight or one time `force` as a binary decision between two tickers does solve the limited collision resolutions requirements like trying to avoid supervision timeouts in connections, periodic sync, BIG Sync and ISO connections. I see `score` in addition to `force` helping here. The decision to pre-empt shall be based on the combination of no. of `lazy_current` above the requested `lazy_periodic`, `score` which is the no. of consecutive `aborts` and `too lates`, and instantaneous applied `force`.  `score` can be reflected into `lazy_current`, if it is just to let ticker know that the timeout callback was skipped in the LLL pipeline.

> did I miss some important fact here?
Not sure. But, my concern is towards the duty cycle of collision resolutions, its complexity, scheduling overheads, and ability to share the time reservations to roles.

If its ok, lets do small incremental improvements/replacements. If its ok, we can start with introducing the concept of `score`, either using explicit `score` variable in ticker node, or translating it into equivalent `lazy_current` value.

mtpr-ot replied:
> These time reservations shall be the bare minimal to guarantee the minimal CPU/Radio and other resource availability.

My main goal of this exercise is to test the theory that we **don't need reservations at all**. The overall concept is that if an event gets to be programmed in the radio, there's a good chance it will finish - but maybe not. The priority takes over the role of reservation, such that a higher priority event may preempt a lower priority event, if they collide.
Actually I think it's only advertising, which has an (almost) precise reservation, and still it reserves space for scan req/rsp and connection ind, which more than often is not needed.

For the concept to work, we cannot even reserve the time for ramping up the radio, there cannot be reservations, because we must make sure that when the preemption timer elapses with a higher priority prepare, it always gets to start.

> The decision to pre-empt shall be based on the combination of no. of `lazy_current` above the requested `lazy_periodic`, `score` which is the no. of consecutive `aborts` and `too lates`, and instantaneous applied `force`. `score` can be reflected into `lazy_current`, if it is just to let ticker know that the timeout callback was skipped in the LLL pipeline.

Yes, I agree - except for `lazy_current`. The lazy parameter of the `lll_prepare_param` will only reflect periodic latency, since `lazy_current` is no longer incremented due to collisions. This also means that for instance the slave prepare is always called, since the decision to skip/queue the prepare is made _after_ the call to `lll_slave_prepare`. This also means that we may not need the `latency` variable that I added (but not yet used) - and that `score` becomes the old `lazy_current`, which again means we could just rename `score` to `lazy`, to stay in legacy terms(?) The idea is that the score/lazy value is added to the (periodic) lazy value from the ticker, and then passed to the prepare_cb in the `prepare_param`. This way, the link layer will "see" the same `lazy` value as for the old ticker.

mtpr-ot commented on subsys/bluetooth/controller/ll_sw/lll.h:
```diff
@@ -155,18 +155,26 @@ struct ull_hdr {
 	void *parent;
 	uint8_t is_stop:1;
+#if defined(CONFIG_BT_CTLR_NEXT_GEN_SCHEDULING)
+	uint8_t score;
```
I will separate the priority calculation and scoring in a separate commit. Here are the basics:

_Scoring_ is used as a term for "earning priority", meaning the higher score an event gets, the more likely it is that it will get to go in the air the next time it's scheduled ("event" here used as the recurring scheduling of an instance of a certain role, e.g. ADV or CONN[0]).

An increase in score is intended to happen whenever an event prepare is either scheduled and pre-empted before it ended, or if it was queued in the prepare pipeline and when popped, was too late to execute (referred to as "aborted" and "too_late").

So when two events "compete" for the same air-time, loser will gain a score of 1 each time the other gets the air, and thus gain higher priority (using its score) the next time, thereby winning the air-time.
This is similar to how the `lazy_current` used to increment when ticker priority resolve caused a loss. Latency is also maintained, in order to be able to supply the correct value in the prepare callback.

mtpr-ot commented on subsys/bluetooth/controller/ticker/ticker.h:
```diff
@@ -152,7 +155,8 @@ uint32_t ticker_job_idle_get(uint8_t instance_index, uint8_t user_id,
 uint32_t ticker_ticks_now_get(void);
 uint32_t ticker_ticks_diff_get(uint32_t ticks_now, uint32_t ticks_old);
-#if !defined(CONFIG_BT_TICKER_COMPATIBILITY_MODE)
+#if !defined(CONFIG_BT_TICKER_COMPATIBILITY_MODE) && \
+	!defined(CONFIG_BT_TICKER_SLOT_AGNOSTIC)
```
OK, but also in the cases where an #ifdef'ed struct member is used in the code, there is no getting around the #if/#endif on that line.

mtpr-ot committed:
Bluetooth: controller: Rename BT_TICKER_COMPATIBILITY_MODE

Renamed to BT_TICKER_LOW_LAT in ticker, and changed selected
conditionally compiled code to use IS_DEFINED macro.

Signed-off-by: Morten Priess <mtpr@oticon.com>

mtpr-ot committed:
Bluetooth: controller: Introducing slot agnostic ticker

Added config BT_TICKER_SLOT_AGNOSTIC which configures the ticker in
a "slot agnostic" mode, in which no collision resolving or slot
reservation is possible or used. In this mode, the ticker acts as a
simple timer.

Signed-off-by: Morten Priess <mtpr@oticon.com>

mtpr-ot committed:
Bluetooth: controller: Propagate ticker force to ULL

The ticker 'force' flag is propagated via the ticker elapsed callback,
in order to provide necessary information for collision resolving in the
link layer.

Signed-off-by: Morten Priess <mtpr@oticon.com>

mtpr-ot committed:
Bluetooth: controller: Move common LLL from vendor file

Split lll_prepare and lll_resume from Nordic LLL to common file for
reuse by all vendors. The split also supports new JIT Scheduling by
defining a common place to calculate event prepare priority.

The module may also house other common parts of the LLL currently
re-implemented identically by vendors.

Signed-off-by: Morten Priess <mtpr@oticon.com>

mtpr-ot committed:
Bluetooth: controller: Clean up LLL priority

Priority in the legacy stack has been unused and "work in progress" for
some time. With this commit, the priority passing/handling is cleaned
up, preparing for the new JIT scheduling priority handling.

Signed-off-by: Morten Priess <mtpr@oticon.com>

mtpr-ot committed:
Bluetooth: controller: Framework for new feature BT_CTLR_JIT_SCHEDULING

This option enables BT_TICKER_SLOT_AGNOSTIC which eliminates
priorities and collision resolving in the ticker.

Event scheduling states are stored in the lll_hdr, and event priority
is passed from LLL implementation, and runtime priority calculated.

LLL implementation decides whether to program radio, start preemption
timer, and/or queue prepare in the prepare pipeline.

Event arbitration is made possible via the common LLL, but not yet
implemented in Nordic LLL.

Signed-off-by: Morten Priess <mtpr@oticon.com>

cvinayak commented on subsys/bluetooth/controller/ticker/ticker.c:
```diff
@@ -922,10 +929,12 @@ void ticker_worker(void *param)
 				/* Reset latency to periodic offset */
 				ticker->lazy_current = 0U;
 				ticker->force = 0U;
+#if !defined(CONFIG_BT_TICKER_SLOT_AGNOSTIC)
```
Please conditionally compile out `ticks_slot` member in the `struct ticker_node` and correspondingly correct the `TICKER_NODE_T_SIZE`

cvinayak commented:
@mtpr-ot If there is no more work in this PR, please resolve conflicts, remove DNM and put the PR to ready, so that we can merge this asap.

mtpr-ot commented:
> @mtpr-ot If there is no more work in this PR, please resolve conflicts, remove DNM and put the PR to ready, so that we can merge this asap.

Will do. I was considering proposing a draft implementation for Nordic LLL, but that can come separately.

mtpr-ot commented:
CI failure seems unrelated. I noticed other PRs also fail with this problem!

mtpr-ot commented:
Close/reopen PR to trigger CI.

mtpr-ot committed:
Bluetooth: controller: LLL refactoring of openisa for API changes

Update openisa LLL to match common LLL changes and make CI pass.

Signed-off-by: Morten Priess <mtpr@oticon.com>

mtpr-ot commented:
Close and reopen to trigger CI.
